docs/sources/source/tutorials/callbacks.rst

```diff
@@ -1,171 +1,182 @@
-=============
-NeMo uses callbacks to do a variety of helper functions during training.
-NeMo comes with three useful callbacks: SimpleLossLoggerCallback,
-CheckpointCallback, and EvaluatorCallback. Callbacks are defined prior to
-calling the train() function, and are passed to the train() function.
-For example, a common training script will look like:
+Callbacks
+=========
+NeMo has a callback system that can be used to inject user code and logic inside its training loop. NeMo's callbacks
+defines the following events that users can inject into:
 
-.. code-block:: python
+.. code-block::
 
-    loggercallback = nemo.core.SimpleLossLoggerCallback(...)
-    savercallback = nemo.core.CheckpointCallback(...)
-    evalcallback = nemo.core.EvaluatorCallback(...)
+    -- on_action_start
+    ---- on_epoch_start
+    ------ on_step_start
+    -------- on_batch_start
+    -------- on_batch_end
+    ------ on_step_end
+    ---- on_epoch_end
+    -- on_action_end
 
-    nf.train(
-        callbacks=[loggercallback, savercallback, evalcallback],
-        ...)
-
-SimpleLossLoggerCallback
-------------------------
-SimpleLossLoggerCallback is used to
-log training metrics such as loss and time per step to screen and to
-tensorboard. SimpleLossLoggerCallback has one required arguments, and two
-arguments that we recommend to overwrite. It requires tensors which is a list
-of NMTensors that print_func(), get_tb_values(), and log_to_tb_func will
-receive as input during
-training. The two reccomended arguments to override are print_func(), and
-either get_tb_values() or log_to_tb_func().
-
-print_func() should be used to log values to screen. We recommend using
-logging.info() in place
-of print(). For example, it can be used to print the loss value:
+At a high level, the NeMo training loop looks like this:
 
 .. code-block:: python
 
-    from nemo import logging
-    def my_print_func(tensors):
-        logging.info(f"Loss {tensors[0]}")
+    def train():
+        # Do initialization of optimizers, amp, ddp, etc
+        callbacks.on_action_start()  # Do all on_action_start functions
+        for epoch in range(num_epochs):  # Or until max_steps
+            callbacks.on_epoch_start()  # Do all on_epoch_start functions
+            batch_counter = 0
+            for data in dataloader:  # Fetch batches of data
+                if batch_counter == 0:
+                    callbacks.on_step_start()
+                callbacks.on_batch_start()
+
+                ...  # Forward and backward pass
+
+                callbacks.on_batch_end()
+                batch_counter += 1
+                if batch_counter == gradient_accumulation_steps:
+                    # By default, gradient_accumulation_steps = 1
+                    # Note this is passed to train() as batches_per_step and sometimes exposed as args.iter_per_step
+                    optimizer.step()
+                    callbacks.update_loss()  # Note, that the loss tensor is only exposed here
```

Comment:
Hey, @blisc what is that "update_loss()" doing? Is that a special _event type_?

And shouldn't it be "loss tensors are"?

Reply:
There is only 1 loss tensor, which is the summation of `tensors_to_optimize` that is passed to train().
update_loss() just does the equivalent of `state["tensors"]["loss"] = final_loss`. It just updates the loss tensor in the tensor dictionary.

Reply:
Wait... this relates to my questions from some time ago... So can you pass more loss tensors to `tensors_to_optimize` or not...? Let's take that discussion offline.

In here more important question is: is `on_update_loss` an event type? If so, then it should be added to that nice list of events that you provided...

Reply:
Yep you can pass as many tensors as you want to `tensors_to_optimize`. But the loss that we call backwards on is the summation of all tensors in `tensors_to_optimize`.

At a very simplified overview it more or less works like this, assuming no gradient accumulation:
```
final_loss = 0
for t in tensors_to_optimize:
    final_loss += t
final_loss.backwards()
optimizer.step()
state["tensors"]["loss"] = final_loss
```

I don't think `on_update_loss` is important as it occurs at exactly the same time as `on_step_end` so it seems redundant.